Improve create tables doc, replication doc, and cluster settings doc (#11319)
crate/crate: docs/general/ddl/replication.rst

.. _ddl-replication:

===========
Replication
===========

You can configure CrateDB to *replicate* tables. When you configure
replication, CrateDB will try to ensure that every table :ref:`shard
<ddl-sharding>` has one or more copies available at all times.

When there are multiple copies of the same shard, CrateDB will mark one as the
*primary shard* and treat the rest as *replica shards*. Write operations
always go to the primary shard, whereas read operations can go to any
shard. CrateDB continually synchronizes data from the primary shard to all
replica shards (through a process known as :ref:`shard recovery
<gloss-shard-recovery>`).

When a primary shard is lost (e.g., due to node failure), CrateDB will promote
a replica shard to primary. Hence, more table replicas mean a smaller chance of
permanent data loss (through increased `data redundancy`) in exchange for more
disk space utilization and intra-cluster network traffic.

CrateDB :ref:`allocates <gloss-shard-allocation>` every shard to a specific

> node. CrateDB will try to allocate the primary shard and replica shards to
> different nodes to maximize :ref:`durability <concept-durability>` and
> :ref:`resiliency <concept-resiliency>`.

Reviewer: ```suggestion
node. CrateDB will enforce the allocation of the primary shard and replica 
shards on different nodes to maximize :ref:`durability <concept-durability>`
and :ref:`resiliency <concept-resiliency>`.
If not enough nodes are available for each replica shard to be allocated,
the table will go into a `YELLOW` health state, see :ref:`sys.health <sys-health>`.
```

Author: ty!!

I reworded slightly::

    CrateDB :ref:`allocates <gloss-shard-allocation>` every shard to a specific
    node. Normally, CrateDB will enforce the requirement that the primary shard and
    replica shards are each allocated to a different node. If that requirement
    cannot be satisfied, the cluster will be given a *yellow* :ref:`health status
    <sys-health>`.

Author: gotcha. thanks!

Replication can also improve read performance because any increase in the
number of shards distributed across a cluster also increases the opportunities
for CrateDB to `parallelize`_ query execution across multiple nodes.

.. TIP::

    The `CrateDB Admin UI`_ provides visual indicators of cluster health that
    take replication status into account. Alternatively, you can query health
    information for yourself directly from the :ref:`sys.health <sys-health>`
    table. You can query detailed replication information from the
    :ref:`sys.shards <sys-shards>` and :ref:`sys.allocations <sys-allocations>`
    tables.

You can configure the number of per-shard replicas :ref:`WITH
<sql-create-table-with>` the :ref:`sql-create-table-number-of-replicas` table
setting.

For example::

    cr> create table my_table10 (
    ...   first_column integer,
    ...   second_column text
    ... ) with (number_of_replicas = 0);
    CREATE OK, 1 row affected (... sec)

> As well as configuring a single fixed number of pre-replica shards, you can use

Reviewer: What do you mean with `pre-replica shards`? 

```suggestion
As well as configuring a single fixed number of replica shards (shard copies), you can use
```

Author: oops! good catch thanks. typo. I think I meant "per-shard replicas" ("per" not "pre" and mixed up the nouns)

I reworded:

> As well as being able to configure a fixed number of replicas, you can configure a range of values by using a string to specify a minimum and a maximum (dependent on the number of nodes in the cluster).

a string to configure a range by specifying a minimum and a maximum (dependent
on the number of nodes in the cluster).

Here are some examples of replica ranges:

========= =====================================================================
Range     Explanation
========= =====================================================================
``0-1``   If you only have one node, CrateDB will not create any replicas. If
          you have more than one node, CreateDB will create one replica per
          shard.

          This range is the default value.
--------- ---------------------------------------------------------------------
``2-4``   Each table will require at least two replicas for CrateDB to consider
          it fully replicated (i.e., a *green status*).

          If the cluster has five nodes, CrateDB will create four replicas,
          with each replica located on a different node from its respective
          primary.

>           If a cluster has four or fewer nodes, CreateDB will have to locate
>           one or more replica shards will be on the same node as the respective
>           primary shard. If this happens, your cluster will have a *yellow*

Reviewer: ```suggestion
          If a cluster has four or fewer nodes, CreateDB would have to locate
          one or more replica shards on the same node as the respective
          primary shard. This is not permitted and such your cluster will have a *yellow*
```

Reviewer: > because even though it's not permitted, it can still happen

No this cannot happen, its really enforced. CrateDB will never allocate a replica shard on the same node of the primary shard.

Reviewer: So we should also remove the "Normally" inside `Normally, CrateDB dynamically allocates shards to...`

Author: ah, I see. thanks for the explanation! I have an outstanding question for you that is related to this. see https://github.com/crate/crate/issues/11263

if you have time, could you take a look at that? based on the discussion there, I think I should probably make a follow-up PR to really make this behavior clear in other places too

.. _ddl-replication:

===========
Replication
===========

You can configure CrateDB to *replicate* tables. When a table is replicated,

> table data is stored on multiple primary shards, with each primary shard having

Author: I have reworked this entire intro. see next comment

one or more replica shards.

When a primary shard is lost (e.g., due to node failure), CrateDB will promote
a replica shard to primary. Hence, more table replicas mean a smaller chance of
permanent data loss (through increased `data redundancy`) in exchange for more
disk space utilization and intra-cluster network traffic.

Replica shards can also improve read performance because more copies of the
data spread across more nodes. Having more copies of the data on more nodes can
also increases the opportunities for CrateDB to `parallelize`_ query

> execution.

Author: based on this and other comments you made, I reworded and expanded the intro:

> You can configure CrateDB to replicate tables. When you configure replication, CrateDB will try to ensure that every table shard has one or more available copies at all times.
>
> When there are multiple copies of the same shard, CrateDB will mark one as the primary shard and treat the rest as replica shards. Write operations always go to the primary shard, whereas read operations can go to any shard. CrateDB continually synchronizes data from the primary shard to all replica shards (through a process known as shard recovery).
>
> When a primary shard is lost (e.g., due to node failure), CrateDB will promote a replica shard to primary. Hence, more table replicas mean a smaller chance of permanent data loss (through increased data redundancy) in exchange for more disk space utilization and intra-cluster network traffic.
> 
> CrateDB allocates every shard to a specific node.CrateDB will try to allocate the primary shard and replica shards to different nodes to maximize durability and resiliency.
>
> Replication can also improve read performance because any increase in the number of shards distributed across a cluster increases the opportunities for CrateDB to parallelize query execution across multiple nodes.

.. TIP::

    The `CrateDB Admin UI`_ provides visual indicators of cluster health that
    take replication status into account. Alternatively, you can query health
    information for yourself directly from the :ref:`sys.health <sys-health>`

>     table.

Author: now reads:

> The CrateDB Admin UI provides visual indicators of cluster health that take replication status into account. Alternatively, you can query health information for yourself directly from the sys.health table. Detailed replication information can be queried from the sys.shards and sys.allocations tables.

You can configure the number of per-shard replicas :ref:`WITH
<sql-create-table-with>` the :ref:`sql-create-table-number-of-replicas` table
setting.

For example::

    cr> create table my_table10 (
    ...   first_column integer,
    ...   second_column text
    ... ) with (number_of_replicas = 0);
    CREATE OK, 1 row affected (... sec)

As well as configuring a single fixed number of pre-replica shards, you can use
a string to configure a range by specifying a minimum and a maximum (dependent
on the number of nodes in the cluster).

Here are some examples of replica ranges:

========= =====================================================================
Range     Explanation
========= =====================================================================
``0-1``   If you only have one node, CrateDB will not create any replicas.

>           Having no replicas will result in a *yellow* cluster health. If you

Author: ah thx for the correction. I copied this from the original text

reworded:

> If you only have one node, CrateDB will not create any replicas. If you have more than one node, CreateDB will create one replica per shard.
>
> This range is the default value.

          have more than one node, CreateDB will create one replica per shard.
--------- ---------------------------------------------------------------------
``2-4``   Each table will require at least two replicas for CrateDB to consider
          it fully replicated (i.e., a *green status*).

>           If the cluster has four nodes, CrateDB will create four replicas,

Author: oops good catch! thx

reworded to fix. I also removed the now redundant note about resiliency as I cover this explicitly in the intro (which is where such an important point belongs imo)

> If the cluster has five nodes, CrateDB will create four replicas, with each replica located on a different node from its respective primary.
>
> If a cluster has four or fewer nodes, CreateDB will have to locate one or more replica shards will be on the same node as the respective primary shard. If this happens, your cluster will have a yellow cluster health.

          with each replica located on a different node from its respective
          primary to improve :ref:`resiliency <concept-resiliency>` (e.g., due
          to node failure).

          If a cluster has three or fewer nodes, CreateDB will have to locate one
          or more  replica shards will be on the same node as the respective
          primary shard. If this happens, your cluster will have a *yellow*

>           cluster health.

Author: fixed as above

--------- ---------------------------------------------------------------------

> ``0-all`` CrateDB will create as many replicas as you have nodes available.

Author: I want to avoid the inline literals if possible. how about this:

> CrateDB will create one replica shard for every node that is available in addition to the node that holds the primary shard.

following:

.. rst-class:: open

- For every lost primary shard, locate a replica and promote it to primary.

  When CrateDB promotes a replica to primary, it can no longer function as a
  replica, and so the total number of replicas decreases by one. Because each
  primary requires a fixed :ref:`sql-create-table-number-of-replicas`, a new
  replica has to be created (see next item).

- For every primary with too few replicas (due to node loss or replica
  promotion), use the primary shard to :ref:`recover <gloss-shard-recovery>`
  the required number of replicas.

Shard recovery is one of the features that allows CrateDB to provide continuous
`availability`_ and `partition tolerance`_ in exchange for some
:ref:`consistency trade-offs <concept-resiliency-consistency>`.

.. SEEALSO::

    `Wikipedia: CAP theorem`_

.. _ddl-replication-underreplication:

Underreplication
================

Having more replicas per primary and distributing shards as thinly as possible
(i.e., fewer shards per node) can both increase chances of a :ref:`successful
recovery <ddl-replication-recovery>` in the event of node loss.

Although not ideal, a single node can hold multiple shards belonging to the

> same table. For example, suppose a table has more shards (primaries and

Reviewer: Having multiple shards of the same table on one node is a common pattern and even recommended as it increases parallel execution per read query (multiple threads can be used).

Author: gotcha thx!
